### ISSUE TITLE
compiler does not catch comparison with "undefined"

### ISSUE THREAD

**Issue by mlawren** [ziglang/zig#8520]
Based on an example from the documentation:

```zig
const expect = @import("std").testing.expect;
const Empty = struct {
    pub const PI = 3.14;
    pub const PI2 = undefined;
};
test "struct namespaced variable" {
    expect(Empty.PI == 3.14);
    expect(Empty.PI2 == 3.14); // fails to fail
    expect(@sizeOf(Empty) == 0);

    // you can still instantiate an empty struct
    const does_nothing = Empty{};
}
```

The above appears to pass a `zig test`.

**Comment by g-w1** [comment]
This is undefined behavior, anything can happen. I think you might be confused with `undefined` in zig and `undefined` in other languages like javascript. In zig, branching on `undefined` means *anything* can happen, in fact llvm probably optimizes it out in this case. The real bug is zig not catching the undefined behavior here. #8529 solves a case close to this, but not exactly.

**Comment by komuw** [comment]
> I think you might be confused with undefined in zig and undefined in other languages like javascript.

Is there an opportunity here for zig to choose another term that doesn't bring with it meaning from other languages?

**Comment by xackus** [comment]
Just to be sure nobody gets confused: `undefined == 3.14` is perfectly legal and the result is undefined.
However, `expect(undefined == 3.14)` branches on the result of the comparison, which is illegal.

**Comment by avdva** [comment]
`Zig` currently disallows comparison between an `int` and `undefined` inside `if` condition, but it is possible to compare a [pointer](https://github.com/ziglang/zig/issues/8056) and `undefined`.
Also, you can do this
```
var x: u8 = 0;
var b = x == undefined;
// or
// const c = undefined;
// b = c == undefined;
if (b) {}
```
which currently compiles fine.
So, the question is: should _any_ `==undefined` comparison produce a compilation error? I was working on a fix for #8056, and didn't find the place where this was explained.

**Comment by andrewrk** [comment]
duplicate of #63
